File: TicTacToe/tictactoe/board.py
class Board:
    def __init__(self, size):
        self.reset(size)

    def reset(self, size):
        self.board = [['' for _ in range(size)] for _ in range(size)]
        self.rowCounts = {}
        self.colCounts = {}
        self.diagCounts = {}
        self.size = size
        self.currCounter = 0
    
    def place(self, player, x, y):
        identifier = player.identifier

        if not (0 <= x < self.size and 0 <= y < self.size):
            raise ValueError("Invalid move: out of bounds")

        if self.board[x][y] != '':
            raise ValueError("Invalid move: cell already occupied.")

        
        self.board[x][y] = identifier

        self.currCounter += 1
        
        #row 
        self.rowCounts[x] = self.rowCounts.get(x, {})
        self.rowCounts[x][identifier] = self.rowCounts[x].get(identifier, 0) + 1

        if self.rowCounts[x][identifier] == self.size:
            return True
            
        #column
        self.colCounts[y] = self.colCounts.get(y, {})
        self.colCounts[y][identifier] = self.colCounts[y].get(identifier, 0) + 1

        if self.colCounts[y][identifier] == self.size:
            return True
            

        #diagonals

        #forward
        if x == y:
            self.diagCounts['forwards'] = self.diagCounts.get('forwards', {})
            self.diagCounts['forwards'][identifier] = self.diagCounts['forwards'].get(identifier, 0) + 1

            if self.diagCounts['forwards'][identifier] == self.size:
                return True
            
        #backward
        if x + y == self.size - 1:
            self.diagCounts['backwards'] = self.diagCounts.get('backwards', {})
            self.diagCounts['backwards'][identifier] = self.diagCounts['backwards'].get(identifier, 0) + 1

            if self.diagCounts['backwards'][identifier] == self.size:
                return True
            
        #tie check
        if self.currCounter == self.size ** 2:
            print("Game Tied!")
            return None
            
        return False

File: TicTacToe/tictactoe-kotlin/Board.kt
class Board constructor(
    private val size: Int
) {

    private lateinit var board: MutableList<MutableList<String>>
    private lateinit var rowCounts: MutableMap<Int, MutableMap<String, Int>>()
    private lateinit var colCounts: MutableMap<Int, MutableMap<String, Int>>()
    private lateinit var diagCounts: MutableMap<String, MutableMap<String, Int>>()
    private val boardSize: Int = size
    private var currCounter: Int = 0

    fun reset(size: Int) {
        board = MutableList(size) {MutableList(size) {""}}
        rowCounts = mutableMapOf<Int, MutableMap<String, Int>>()
        colCounts = mutableMapOf<Int, MutableMap<String, Int>>()
        diagCounts = mutableMapOf<String, MutableMap<String, Int>>()
        currCounter = 0
    }

    fun place(player: Player, x: Int, y: Int): GameResult {
        val identifier = player.identifier

        if (!(x >= 0 && x < boardSize)||!(y >= 0 && y < boardSize)) {
            throw IllegalArgumentException("Invalid move: out of bounds")
        }

        if (board[x][y] != "") {
            throw IllegalArgumentException("Invalid move: cell already occupied")
        }

        board[x][y] = identifier
        currCounter += 1

        //row
        val rowInner = rowCounts.getOrPut(x) { mutableMapOf() }
        rowInner[identifier] = (rowInner[identifier] ?: 0) + 1

        if (rowCounts[x]!![identifier]!! == boardSize) {
            return GameResult.Winner(player)
        }

        //column
        val colInner = colCounts.getOrPut((y)) { mutableMapOf() }
        colInner[identifier] = (colInner[identifier] ?: 0) + 1
        
        if (colCounts[y]!![identifier]!! == boardSize) {
            return GameResult.Winner(player)
        }

        //diagonal

        //forwards
        if (x == y) {
            val forwardDiagInner = diagCounts.getOrPut("forwards") { mutableMapOf() }
            forwardDiagInner[identifier] = (forwardDiagInner[identifier] ?: 0) + 1

            if (diagCounts["forwards"]!![identifier]!! == boardSize) {
                return GameResult.Winner(player)
            }
        }

        //backwards

        if (x + y == boardSize - 1) {
            val backwardsDiagInner = diagCounts.getOrPut("backwards") { mutableMapOf() }
            backwardsDiagInner[identifier] = (backwardsDiagInner[identifier] ?: 0) + 1

            if (diagCounts["backwards"]!![identifier]!! == boardSize) {
                return GameResult.Winner(player)
            }
        }

        //tie check

        if (currCounter == boardSize * boardSize) {
            println("Game Tied")
            return GameResult.Tie
        }
        return GameResult.Ongoing
    }
}



